File: ios/MobiCore/Core/EmulatorEngine.swift
import Combine
import CoreGraphics
import Foundation
import CoreMotion
import GameController
import QuartzCore
import SwiftUI

/// Điều khiển một trò chơi đang chạy.
///
/// A J2ME game loop blocks and sleeps, so the MIDlet runs on a background
/// queue; only the finished frame is handed to the main actor. The display
/// link paces presentation so the UI redraws at most once per screen refresh
/// no matter how fast the game runs.
@MainActor
final class EmulatorEngine: ObservableObject {

    @Published private(set) var frame: CGImage?
    @Published private(set) var isRunning = false
    @Published private(set) var isPaused = false
    @Published private(set) var error: String?
    /// Vì sao game vừa chết, khi nó chết. Xem `CrashReading`.
    @Published private(set) var crash: CrashReading?
    @Published private(set) var screenSize = CGSize(width: 240, height: 320)
    /// Labels the running screen has mapped to the two softkeys.
    @Published private(set) var leftSoftKeyLabel: String?
    @Published private(set) var rightSoftKeyLabel: String?
    /// True while the emulated screen draws the command bar itself, which on
    /// a touchscreen is the softkeys: tapping a label runs its command.
    @Published private(set) var showsSoftKeyBar = false

    /// How solid the keypad should be drawn now, in percent.
    ///
    /// Read off the emulator on every presented frame rather than worked out
    /// here: the session holds both the setting and the clock that fades it.
    @Published private(set) var keypadOpacity = 100
    /// How fast the game is playing, as a percentage of a handset's pace.
    @Published private(set) var speed = 100

    private let bridge = MobiCoreBridge.shared
    private let queue = DispatchQueue(label: "com.mobicore.midlet", qos: .userInitiated)
    private var displayLink: CADisplayLink?
    private var running = false
    private var frameLimit = 30

    func start(suiteId: String, settings: GameSettings?) {
        stop()
        // Starts the game and, if the player left one behind, puts it back
        // where it was.
        let response = bridge.startGameResuming(suiteId)
        guard let result: ActionResult = decode(response), result.ok else {
            error = decode(response).flatMap { (r: ActionResult) in r.error } ?? "Không khởi động được trò chơi"
            return
        }
        error = nil
        crash = nil
        screenSize = CGSize(width: result.width ?? 240, height: result.height ?? 320)
        frameLimit = max(0, settings?.frameLimit ?? 30)
        isRunning = true
        isPaused = false
        running = true

        let link = CADisplayLink(target: self, selector: #selector(present))
        link.preferredFramesPerSecond = frameLimit == 0 ? 0 : min(60, frameLimit)
        link.add(to: .main, forMode: .common)
        displayLink = link

        refreshSoftKeys()
        queue.async { [weak self] in
            self?.runLoop()
        }
    }

    private nonisolated func runLoop() {
        while DispatchQueue.main.sync(execute: { self.running }) {
            let started = CACurrentMediaTime()
            // The frame budget follows the speed control: at double speed the
            // game's own clock runs twice as fast, and drawing at the old rate
            // would show half of what it does.
            let interval: TimeInterval = DispatchQueue.main.sync {
                self.frameLimit > 0
                    ? 1.0 / (Double(self.frameLimit) * Double(max(10, self.speed)) / 100.0)
                    : 0
            }
            let paused = DispatchQueue.main.sync { self.isPaused }
            if !paused {
                _ = bridge.renderFrame()
            }
            if bridge.isFinished {
                DispatchQueue.main.async { self.finish() }
                return
            }
            let elapsed = CACurrentMediaTime() - started
            let remaining = interval - elapsed
            Thread.sleep(forTimeInterval: remaining > 0 ? remaining : 0.001)
        }
    }

    /// Pulls the newest frame on the display's own cadence.
    @objc private func present() {
        guard running, !isPaused else { return }
        if let image = bridge.copyFrameImage()?.takeRetainedValue() {
            frame = image
        }
        refreshTextInput()
        let solidity = Int(bridge.keypadDrawOpacity())
        if solidity != keypadOpacity {
            keypadOpacity = solidity
        }
        // Hỏi bằng một cờ chứ không dựng JSON mỗi khung hình: phần lớn thời
        // gian câu trả lời là "không có gì", và chuỗi ấy sẽ bị vứt đi ngay.
        if crash == nil, bridge.hasCrashed {
            let reading: CrashReading? = decode(bridge.crashJSON())
            if let reading, reading.has {
                crash = reading
                // Game đã chết thì vòng lặp không còn gì để chạy, nhưng phiên
                // vẫn giữ nguyên để người chơi còn thấy màn hình cuối.
                running = false
                isRunning = false
            }
        }
    }

    private func finish() {
        stop()
    }

    /// Người chơi đã đọc xong lời báo hỏng; game đã chết được dọn đi.
    func dismissCrash() {
        crash = nil
        error = nil
        _ = bridge.dismissCrash()
        stop()
    }

    func pause() {
        guard isRunning, !isPaused else { return }
        isPaused = true
        bridge.pauseGame()
    }

    func resume() {
        guard isRunning, isPaused else { return }
        isPaused = false
        bridge.resumeGame()
    }

    func stop() {
        running = false
        // Một game kẹt trong vòng lặp của chính nó không bao giờ đọc tới cờ
        // dừng ở trên: lệnh này xuyên thẳng vào máy ảo, nên rời một game treo
        // là chuyện tức thì chứ không phải chờ hết giờ.
        _ = bridge.requestStopGame()
        displayLink?.invalidate()
        displayLink = nil
        if isRunning {
            // Leaving a game saves it: on a phone, leaving is not always the
            // player's decision.
            bridge.stopGameSaving()
        }
        isRunning = false
        isPaused = false
    }

    // MARK: - Text entry

    /// True while the game is asking for text; the view raises the keyboard.
    @Published private(set) var wantsText = false
    @Published var text = ""

    /// Called each frame: the game decides when it wants text, not the view.
    private func refreshTextInput() {
        let active = bridge.isTextInputActive()
        if active != wantsText {
            wantsText = active
            text = active ? bridge.textInput() : ""
        }
    }

    /// Whole strings rather than key events: the keyboard does its own
    /// editing, and what the game should see is the result.
    func commitText(_ value: String) {
        bridge.setTextInput(value)
    }

    // ------------------------------------------------------------- tilting

    /// The phone's own idea of which way it is being held.
    private let motion = CMMotionManager()

    /**
     Starts reading how the phone is held.

     What the emulator wants is "how far is this leaning", from -1 to 1, and
     Core Motion's gravity vector is already in those units: a phone on its
     side reads 1 on one axis. Held upright in portrait, gravity is along -y,
     so leaning right tips it into x and leaning away tips it out of y.

     Reading stops with the game. A sensor left running is a battery drained
     for a screen nobody is looking at.
     */
    func startTilting() {
        guard motion.isDeviceMotionAvailable, !motion.isDeviceMotionActive else { return }
        // The game reads held keys at its own pace, so a faster rate would be
        // readings nobody looks at.
        motion.deviceMotionUpdateInterval = 1.0 / 30
        motion.startDeviceMotionUpdates(to: .main) { [weak self] data, _ in
            guard let gravity = data?.gravity else { return }
            let x = max(-1, min(1, gravity.x))
            let y = max(-1, min(1, -gravity.y - 1))
            _ = self?.bridge.tiltedX(Int(x * 1000), y: Int(y * 1000))
        }
    }

    func stopTilting() {
        if motion.isDeviceMotionActive {
            motion.stopDeviceMotionUpdates()
        }
    }

    // ---------------------------------------------------------- the controller

    /// True while a controller is connected, so the game can say so.
    @Published private(set) var padConnected = false

    /// Which directions the stick is pushing, so a change can be told apart
    /// from the stream of positions a stick actually sends.
    private var stickHeld: Set<String> = []

    /**
     Starts listening for a controller.

     iOS hands out controllers through notifications rather than events, so
     this watches for one arriving and binds its buttons when it does. Bound
     to the emulator's own names — `padA`, `padUp` — because what those do is
     the profile's decision, made once for iOS and Android alike.
     */
    func watchForControllers() {
        NotificationCenter.default.addObserver(
            forName: .GCControllerDidConnect, object: nil, queue: .main
        ) { [weak self] note in
            guard let controller = note.object as? GCController else { return }
            Task { @MainActor in self?.bind(controller) }
        }
        NotificationCenter.default.addObserver(
            forName: .GCControllerDidDisconnect, object: nil, queue: .main
        ) { [weak self] _ in
            Task { @MainActor in
                self?.padConnected = !GCController.controllers().isEmpty
            }
        }
        for controller in GCController.controllers() {
            bind(controller)
        }
    }

    private func bind(_ controller: GCController) {
        guard let pad = controller.extendedGamepad else { return }
        padConnected = true

        func hold(_ name: String) -> GCControllerButtonValueChangedHandler {
            { [weak self] _, _, pressed in
                Task { @MainActor in
                    if pressed {
                        self?.pressPad(name)
                    } else {
                        self?.releasePad(name)
                    }
                }
            }
        }

        pad.buttonA.pressedChangedHandler = hold("padA")
        pad.buttonB.pressedChangedHandler = hold("padB")
        pad.buttonX.pressedChangedHandler = hold("padX")
        pad.buttonY.pressedChangedHandler = hold("padY")
        pad.leftShoulder.pressedChangedHandler = hold("padL1")
        pad.rightShoulder.pressedChangedHandler = hold("padR1")
        pad.leftTrigger.pressedChangedHandler = hold("padL2")
        pad.rightTrigger.pressedChangedHandler = hold("padR2")
        pad.buttonMenu.pressedChangedHandler = hold("padStart")
        pad.buttonOptions?.pressedChangedHandler = hold("padSelect")

        pad.dpad.up.pressedChangedHandler = hold("padUp")
        pad.dpad.down.pressedChangedHandler = hold("padDown")
        pad.dpad.left.pressedChangedHandler = hold("padLeft")
        pad.dpad.right.pressedChangedHandler = hold("padRight")

        // A J2ME game has four directions and nothing else, so the stick is
        // read as a d-pad. The dead zone is wide on purpose: a worn stick
        // resting at 0.2 would otherwise walk the player into a wall for as
        // long as the game is open.
        //
        // A direction is taken at one angle and given up at a smaller one,
        // the same way tilting works: with a single threshold, a stick
        // resting right on it sends press, release, press, release many times
        // a second, which the game reads as a player hammering the key.
        pad.leftThumbstick.valueChangedHandler = { [weak self] _, x, y in
            Task { @MainActor in
                guard let self else { return }
                let was = self.stickHeld
                func amount(_ button: String, _ pushed: Float) -> Bool {
                    pushed >= (was.contains(button) ? 0.35 : 0.5)
                }
                var held: Set<String> = []
                if amount("padLeft", -x) { held.insert("padLeft") }
                if amount("padRight", x) { held.insert("padRight") }
                if amount("padUp", y) { held.insert("padUp") }
                if amount("padDown", -y) { held.insert("padDown") }
                self.stickMoved(held)
            }
        }
    }

    /// A control on a real pad was pressed. The profile decides what it does.
    func pressPad(_ pad: String) {
        _ = bridge.pressPad(pad)
        refreshSoftKeys()
    }

    func releasePad(_ pad: String) {
        _ = bridge.releasePad(pad)
    }

    /// What the stick is pushing now, as presses and releases.
    ///
    /// A stick sends positions, not presses, so the change is worked out here.
    /// Without it a game reading held keys would see one press and then
    /// nothing.
    private func stickMoved(_ directions: Set<String>) {
        guard directions != stickHeld else { return }
        for gone in stickHeld.subtracting(directions) {
            releasePad(gone)
        }
        for fresh in directions.subtracting(stickHeld) {
            pressPad(fresh)
        }
        stickHeld = directions
    }

    func press(_ button: String) {
        bridge.press(button)
        if button == "softLeft" || button == "softRight" {
            // A command may have swapped the screen, and with it the labels.
            refreshSoftKeys()
        }
    }

    /// What the two softkeys carry, and whether the screen shows them itself.
    private struct SoftKeys: Decodable {
        let left: String?
        let right: String?
        let bar: Bool?
    }

    /// Re-reads the softkey labels from the running screen.
    func refreshSoftKeys() {
        guard let data = bridge.softKeysJSON().data(using: .utf8),
              let labels = try? JSONDecoder().decode(SoftKeys.self, from: data) else {
            leftSoftKeyLabel = nil
            rightSoftKeyLabel = nil
            showsSoftKeyBar = false
            return
        }
        leftSoftKeyLabel = labels.left
        rightSoftKeyLabel = labels.right
        showsSoftKeyBar = labels.bar ?? false
    }

    /// A J2ME game paces itself off the clock, so this changes what it is
    /// told the time is; the game does the rest with its own logic intact.
    func cycleSpeed() {
        guard let data = bridge.cycleSpeed().data(using: .utf8),
              let payload = try? JSONDecoder().decode(SpeedResponse.self, from: data) else {
            return
        }
        speed = payload.speed
        // Drawing at the old rate would show half of what the game does.
        displayLink?.preferredFramesPerSecond = frameLimit == 0
            ? 0
            : min(60, max(1, frameLimit * speed / 100))
    }

    private struct SpeedResponse: Decodable {
        let speed: Int
    }

    /// How many seconds of play can still be taken back.
    @Published private(set) var rewindDepth = 0

    /// Takes back the last second or so. These games restart a level on one
    /// mistake, because a handset had nowhere to keep anything else.
    @discardableResult
    func rewind() -> Bool {
        guard let data = bridge.rewindStep().data(using: .utf8),
              let payload = try? JSONDecoder().decode(RewindResponse.self, from: data) else {
            return false
        }
        rewindDepth = payload.seconds
        return payload.ok
    }

    func refreshRewind() {
        guard let data = bridge.rewindJSON().data(using: .utf8),
              let payload = try? JSONDecoder().decode(RewindResponse.self, from: data) else {
            return
        }
        rewindDepth = payload.seconds
    }

    private struct RewindResponse: Decodable {
        let ok: Bool
        let seconds: Int
    }

    func release(_ button: String) {
        bridge.release(button)
    }

    /// Maps a point in the displayed rectangle back to emulated coordinates.
    func pointerDown(at point: CGPoint, in rect: CGRect) {
        guard let mapped = map(point, in: rect) else { return }
        bridge.pointerDown(atX: mapped.x, y: mapped.y)
    }

    func pointerMoved(to point: CGPoint, in rect: CGRect) {
        guard let mapped = map(point, in: rect) else { return }
        bridge.pointerMoved(toX: mapped.x, y: mapped.y)
    }

    func pointerUp(at point: CGPoint, in rect: CGRect) {
        guard let mapped = map(point, in: rect) else { return }
        bridge.pointerUp(atX: mapped.x, y: mapped.y)
    }

    private func map(_ point: CGPoint, in rect: CGRect) -> (x: Int, y: Int)? {
        guard rect.width > 0, rect.height > 0 else { return nil }
        let x = Int((point.x - rect.minX) / rect.width * screenSize.width)
        let y = Int((point.y - rect.minY) / rect.height * screenSize.height)
        guard x >= 0, y >= 0, x < Int(screenSize.width), y < Int(screenSize.height) else { return nil }
        return (x, y)
    }

    func screenshot() -> Data? {
        bridge.screenshotPNG()
    }

    func logLines() -> [String] {
        bridge.logText().split(separator: "\n").map(String.init)
    }

    private nonisolated func decode<T: Decodable>(_ json: String) -> T? {
        guard let data = json.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }
}

File: android/app/src/main/kotlin/com/mobicore/app/emu/EmulatorEngine.kt
package com.mobicore.app.emu

import android.graphics.Bitmap
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.mobicore.core.emu.CrashDiagnosis
import com.mobicore.core.emu.EmulatorSession
import com.mobicore.core.emu.SaveState
import com.mobicore.core.gfx.PngWriter
import com.mobicore.core.jar.SuiteLoader
import com.mobicore.core.model.GameProfile
import com.mobicore.core.storage.LocalVfs
import com.mobicore.core.storage.StorageLayout
import com.mobicore.core.vm.VmCancelled
import com.mobicore.core.vm.VmError
import com.mobicore.core.vm.VmThrow
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Drives one running game on Android.
 *
 * The MIDlet runs on its own thread rather than on the UI thread: a J2ME game
 * loop blocks and sleeps freely, and letting it do that on the main thread
 * would freeze the whole app. Compose is woken by a frame counter, and the
 * bitmap is only touched under [frameLock].
 */
class EmulatorEngine(
    private val filesDir: String,
    /** The phone's motor, when the app has one to offer. */
    private val vibration: com.mobicore.core.haptics.VibrationSink? = null,
) {

    /** Bumped after each painted frame so Compose knows to redraw. */
    var frameCounter by mutableIntStateOf(0)
        private set

    var running by mutableStateOf(false)
        private set

    var paused by mutableStateOf(false)
        private set

    /**
     * Lần hỏng gần nhất, đã đọc thành lời.
     *
     * Trước đây chỗ này là một dòng chữ đỏ ghi tên lớp ngoại lệ — đúng với
     * người viết máy ảo và vô nghĩa với người đang chơi. Bản đọc hiểu nói hỏng
     * cái gì, vì sao và làm gì tiếp; tên ngoại lệ vẫn còn, trong phần kỹ thuật.
     */
    var crash by mutableStateOf<CrashDiagnosis?>(null)
        private set

    /** Ngăn xếp lúc chết, để gửi kèm báo lỗi. */
    var crashStack by mutableStateOf("")
        private set

    var lastError by mutableStateOf<String?>(null)
        private set

    /** Bumped whenever the running screen's commands change. */
    var commandRevision by mutableIntStateOf(0)
        private set

    val frameLock = Any()

    var bitmap: Bitmap? = null
        private set

    private var session: EmulatorSession? = null
    private val audio = AudioTrackSink()
    private var loop: Thread? = null
    private val stopRequested = AtomicBoolean(false)
    private var pixelBuffer: IntArray = IntArray(0)

    fun session(): EmulatorSession? = session

    fun screenWidth(): Int = session?.screen()?.width() ?: 0

    fun screenHeight(): Int = session?.screen()?.height() ?: 0

    /** Boots a suite and starts the game loop. */
    fun start(suite: SuiteLoader, profile: GameProfile) {
        stop()
        lastError = null
        crash = null
        crashStack = ""
        val layout = StorageLayout(StorageLayout.join(filesDir, "MobiCore"))
        val created = EmulatorSession.create(suite, profile, LocalVfs(), layout, AndroidHost())
        // Sound goes to the device rather than to the recorder the core
        // defaults to; the profile's volume is already applied inside the
        // emulator, so the track only has to play what it is handed.
        created.setAudio(audio)
        // Múi giờ của chiếc máy này, giờ mùa hè đã tính sẵn trong độ lệch.
        // Phần lõi không mang bảng múi giờ của thế giới, nên chỗ biết thì nói.
        val zone = java.util.TimeZone.getDefault()
        created.vm().setTimeZone(zone.id, zone.getOffset(System.currentTimeMillis()))
        // A J2ME game's only physical feedback was the handset shaking.
        vibration?.let { created.setVibration(it) }
        session = created
        val width = created.screen().width()
        val height = created.screen().height()
        pixelBuffer = IntArray(width * height)
        bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)

        stopRequested.set(false)
        running = true
        paused = false
        loop = Thread({ runLoop(created, profile) }, "mobicore-midlet").apply {
            isDaemon = true
            priority = Thread.NORM_PRIORITY + 1
            start()
        }
    }

    private fun suiteHolds(session: EmulatorSession, className: String): Boolean =
        session.info().midlets().any { it.className() == className }

    private fun runLoop(active: EmulatorSession, profile: GameProfile) {
        try {
            // The MIDlet the player chose, when the suite still holds it: a
            // JAR often carries a help screen and a settings screen beside
            // the game, and a stale name must not leave the game unopenable.
            val wanted = profile.midletClass()
            if (wanted.isNotEmpty() && suiteHolds(active, wanted)) {
                active.start(wanted)
            } else {
                active.start()
            }
            val limit = profile.frameLimit()

            while (!stopRequested.get() && !active.isFinished) {
                val frameStart = System.nanoTime()
                if (!paused && active.renderFrame()) {
                    publish(active)
                }
                // The frame budget follows the speed control: at double speed
                // the game's own clock runs twice as fast, and drawing at the
                // old rate would show half of what it does.
                val frameNanos = if (limit > 0) {
                    1_000_000_000L * 100L / (limit.toLong() * active.speed().coerceAtLeast(10))
                } else {
                    0L
                }
                if (frameNanos > 0) {
                    val elapsed = System.nanoTime() - frameStart
                    val remaining = frameNanos - elapsed
                    if (remaining > 0) {
                        Thread.sleep(remaining / 1_000_000L, (remaining % 1_000_000L).toInt())
                    }
                } else {
                    // Uncapped still yields, so the UI thread is never starved.
                    Thread.sleep(1)
                }
            }
        } catch (interrupted: InterruptedException) {
            Thread.currentThread().interrupt()
        } catch (thrown: VmThrow) {
            noteCrash(active, thrown)
        } catch (stopped: VmCancelled) {
            // Người chơi dừng, không phải game hỏng: không có gì để báo.
        } catch (error: VmError) {
            noteCrash(active, error)
        } catch (unexpected: RuntimeException) {
            noteCrash(active, unexpected)
        } finally {
            running = false
        }
    }

    /**
     * Ghi lại một lần hỏng, kèm chỗ nó chết.
     *
     * Ngăn xếp phải lấy ngay tại đây: phiên chạy sắp bị dọn, và câu hỏi "chết
     * ở đâu" chỉ trả lời được khi nó còn.
     */
    private fun noteCrash(active: EmulatorSession, failure: Throwable) {
        val read = CrashDiagnosis.of(failure)
        crash = read
        crashStack = active.vm().interpreter().crashTrace()
        lastError = read.reason()
    }

    /** Người chơi đã đọc xong lời báo hỏng. */
    fun dismissCrash() {
        crash = null
        crashStack = ""
        lastError = null
    }

    /** Copies the emulated framebuffer into the bitmap Compose draws. */
    private fun publish(active: EmulatorSession) {
        val target = bitmap ?: return
        val pixels = active.screen().pixels()
        synchronized(frameLock) {
            System.arraycopy(pixels, 0, pixelBuffer, 0, pixelBuffer.size)
            target.setPixels(pixelBuffer, 0, target.width, 0, 0, target.width, target.height)
        }
        frameCounter++
    }

    fun pause() {
        val active = session ?: return
        if (!paused) {
            paused = true
            active.pause()
        }
    }

    fun resume() {
        val active = session ?: return
        if (paused) {
            paused = false
            active.resume()
        }
    }

    // ------------------------------------------------------------ text entry

    /** True while the game is showing a TextBox or a focused text field. */
    fun isTextInputActive(): Boolean = session?.isTextInputActive ?: false

    /** What that field holds now, so the keyboard opens on it. */
    fun textInput(): String = session?.textInput() ?: ""

    /** Puts what the phone's keyboard produced into the field. */
    fun setTextInput(value: String) {
        session?.setTextInput(value)
    }

    /**
     * Captures the running game.
     *
     * Returns null when the game holds something that cannot be written down
     * — an open connection, a sound being played. The caller carries on: not
     * being able to save a position is a disappointment, and refusing to
     * close the game over it would be worse.
     */
    fun captureState(): ByteArray? {
        val active = session ?: return null
        return runCatching { SaveState.capture(active) }.getOrNull()
    }

    /** Puts a captured game back. The session must already be started. */
    fun restoreState(state: ByteArray): Boolean {
        val active = session ?: return false
        return runCatching {
            SaveState.restore(active, state)
            publish(active)
            true
        }.getOrDefault(false)
    }

    /** The screen as it stands, as PNG bytes, for the saved state's picture. */
    fun screenshot(): ByteArray? {
        val active = session ?: return null
        return runCatching { PngWriter.encode(active.screen()) }.getOrNull()
    }

    // ------------------------------------------------------------- gamepad

    /**
     * A control on a real pad was pressed.
     *
     * The name is the emulator's own; the profile decides what it does, so a
     * remapped pad is remapped for the phone and the preview alike.
     *
     * @return true when the control was bound to something
     */
    fun pressPad(pad: String): Boolean {
        val active = session ?: return false
        if (active.profile().gamepad().buttonFor(pad).isEmpty()) {
            return false
        }
        active.pressPad(pad)
        commandRevision++
        return true
    }

    fun releasePad(pad: String) {
        session?.releasePad(pad)
    }

    /** Directions a stick is pushing, so a change can be told from a repeat. */
    var stickHeld = emptySet<String>()
        private set

    /**
     * A stick moved.
     *
     * Android reports a stick as a stream of positions, not as presses, so the
     * change is worked out here: what is newly pushed gets pressed, what is no
     * longer pushed gets released. Without that a game reading held keys would
     * see one press and then nothing.
     */
    fun stickMoved(directions: Set<String>) {
        if (directions == stickHeld) {
            return
        }
        for (gone in stickHeld - directions) {
            releasePad(gone)
        }
        for (fresh in directions - stickHeld) {
            pressPad(fresh)
        }
        stickHeld = directions
    }

    /**
     * The phone was tilted.
     *
     * A sensor reports a position many times a second, not a press; the
     * session works out the change and holds or lets go of the directions.
     */
    fun tilted(x: Float, y: Float) {
        session?.tilted(x, y)
    }

    /** Lets go of whatever tilting was holding, for a game being left. */
    fun releaseTilt() {
        session?.releaseTilt()
    }

    fun stop() {
        stopRequested.set(true)
        // Một game kẹt trong vòng lặp của chính nó không bao giờ đọc tới cờ
        // dừng ở trên: lệnh này xuyên thẳng vào máy ảo, nên rời một game treo
        // là chuyện tức thì chứ không phải chờ hết giờ.
        session?.requestStop()
        loop?.interrupt()
        loop?.join(750)
        loop = null
        session?.destroy()
        session = null
        audio.releaseAll()
        running = false
        paused = false
    }

    /** Restarts the current suite from scratch, discarding VM state only. */
    fun restart(suite: SuiteLoader, profile: GameProfile) {
        stop()
        start(suite, profile)
    }

    fun pressButton(button: String) {
        session?.pressButton(button)
        // Any key can swap the screen, and with it the softkey labels: picking
        // a row on a List opens the next screen just as a softkey command does,
        // and the labels have to follow.
        commandRevision++
    }

    fun releaseButton(button: String) {
        session?.releaseButton(button)
    }

    /**
     * How solid the keypad should be drawn right now, in percent.
     *
     * The session keeps the clock and the profile holds the setting, so the
     * answer comes from there rather than being worked out again here.
     */
    fun keypadOpacity(): Int = session?.keypadOpacity() ?: 100

    /** Brings a faded keypad back, for a touch that is not a key press. */
    fun noteKeypadUse() {
        session?.noteKeypadUse()
    }

    fun pointerDown(x: Int, y: Int) {
        session?.pointerPressed(x, y)
    }

    fun pointerMove(x: Int, y: Int) {
        session?.pointerDragged(x, y)
    }

    fun pointerUp(x: Int, y: Int) {
        session?.pointerReleased(x, y)
    }

    /**
     * Labels the two softkeys should show. A handset leaves them blank until a
     * MIDlet registers a Command, and shows whatever it registered.
     */
    /**
     * True while the emulated screen draws the command bar itself.
     *
     * A touchscreen ran these games with that bar as the button, so while it
     * is there the keypad leaves its two softkeys out rather than repeating
     * the same two words underneath.
     */
    fun showsSoftKeyBar(): Boolean = session?.showsSoftKeyBar() ?: false

    fun leftSoftKeyLabel(): String? = session?.leftSoftKeyLabel()

    fun rightSoftKeyLabel(): String? = session?.rightSoftKeyLabel()

    fun logLines(): List<String> =
        session?.log()?.entries()?.map { it.toString() } ?: emptyList()
}
